app/controllers/alchemy/admin/pages_controller.rb

```diff
@@ -25,6 +27,59 @@ def index
         end
       end
 
+      # Returns all pages as a tree from the root given by the id parameter
+      #
+      def tree
+        authorize! :tree, :alchemy_admin_pages
+
+        tree = []
+        path = [{id: @page.parent_id, children: tree}]
+        page_list = @page.self_and_descendants
+
+        page_list.each_with_index do |page, i|
+          if page.parent_id != path.last[:id]
+            if path.map { |o| o[:id] }.include?(page.parent_id) # Lower level
+              path.pop while path.last[:id] != page.parent_id
+            else # One level up
+              path << path.last[:children].last
+            end
+          end
+
+          has_children = page_list[i + 1] && page_list[i + 1].parent_id == page.id
+          level = path.count
+
+          path.last[:children] << {
```

Comment:
Please have a look into the already present page serializer. We could add additional information needed for the tree conditionally for admin views and if the user has the permission to render the admin page tree.